### ISSUE TITLE
Simplify VarHi expression

### ISSUE THREAD

**Issue by ljwolf** [jeffcsauer/GSOC2020#3]
```python
VarHi = ((n-1)**-1) * \
                     (denom**-2) * \
                     ((np.sum(yresid**2)/n) - yresid_mean**2) * \
                     ((n*np.array([np.sum(np.array(list(w[y].values()))**2) for y in range(len(y))])) - np.array(Wrs)**2)
```
If I'm understanding this correctly, I think this can be re-written in a bit clearer fashion as:
```python
( (n - 1)**-1 * # 1 / (n - 1) 
  (denom **-2) *  # mean of residuals times row sums
  (((yresid**2).mean() - yresid_mean**2) * # E[e**2] - E[e]**2
  (numpy.asarray((w.sparse **2).sum(axis=1)).flatten() - Wrs ** 2) # E[w**2] - E[w]**2
)
```
I think this is clearer only because it removes the intermediate sum of squared weights values. It may also be good to do that separately, for explicitness (no harm in intemediate variables in Python)
```python
squared_rowsum = numpy.asarray((w.sparse**2).sum(axis=1)).flatten()
( (n - 1)**-1 * # 1 / (n - 1) 
  (denom **-2) *  # mean of residuals times row sums
  (((yresid**2).mean() - yresid_mean**2) * # E[e**2] - E[e]**2
  (squared_rowsum - row_sum ** 2) # E[w**2] - E[w]**2
)
```

**Comment by jeffcsauer** [comment]
This makes sense to me and is definitely much clearer. A quick clarification - it should be the sum of each squared row weight, not the square of the sum of the row weights, correct? 

Ideally I would like to simply square the `w.weight.values()` but they are in a dict so it becomes a little tricky. Do you know of another shortcut to square and then sum the following? ideally (but not working):

`np.sum(wq.weights.values()**2)`

My current 'happy medium' is: 
 
```
squared_rowsum = (n*np.array([np.sum(np.array(list(w[i].values()))**2) for i in w.id_order]))

VarHi = ( (n-1)**-1) * \
               (denom**-2) * \
               ((np.sum(yresid**2)/n) - yresid_mean**2) * \
               (squared_rowsum - (rowsum**2)
           )
```

This fixes the previously confusing use of `y` as the iterator and defers to `w.id_order`. Addressed in [commit b720cf1](https://github.com/jeffcsauer/GSOC2020/commit/b720cf1d882197e6fc639df92694e2fba693ffba).

**Comment by ljwolf** [comment]
Great!

> Ideally, I would like to simply square the w.weight.values()

Right, which is why I used `(w.sparse**2).sum(axis=1)` in that last comment. 

That:
1. squares the weight values
2. takes the sum of those values over rows. 
`w.sparse` is a matrix representation of the `w`, and is way faster/more efficient than iterating over the `w.weights` or `w.neighbors` components. 

Where possible, when doing math on the `w`, try to avoid iteration & instead use the `w.sparse` attribute for matrix stuff.

**Comment by jeffcsauer** [comment]
Apologies, my last comment wasn't clear and I should have included a reproducible example!

Loading in some trial data:
```
import libpysal
import geopandas
import numpy as np
df = geopandas.read_file('https://github.com/jeffcsauer/GSOC2020/raw/master/validation/data/neighborhoods.gpkg')
w = libpysal.weights.Rook.from_dataframe(df)
```

If I apply `(w.sparse**2).sum(axis=1)` on non row-standardized weights, this produces a set of values...
```
matrix([[14.],
        [14.],
        [21.],
        .....
        [22.],
        [10.],
        [25.]])
```
 
But to my understanding these are not the values we want.  For example, `w[0]` has three neighbors, and we want the sum of each row weight squared, i.e `1**2 + 1**2 + 1**2 = 3`. 

The existing list comprehension method does produce the correct values...
```
[np.sum(np.array(list(w[i].values()))**2) for i in w.id_order]
```
but I be missing something with the use of sparse weights here? 

Same goes for row-standardized weights.
```
w.transform = 'r'
# Option 1
print((w.sparse**2).sum(axis=1))
# Option 2
print([np.sum(np.array(list(w[i].values()))**2) for i in w.id_order])
```

My implementation draws on the `spdep` implementation of [LOSH, specifically line 31](https://github.com/r-spatial/spdep/blob/be6d0dbb6113909480b624ac0d6a59524b1eddb8/R/LOSH.R#L31) with the `vapply(...)` function .

I thought yours would work and I moved the parts around trying to get it to match, but haven't had any luck. Perhaps it's the placement of `**2`?

**Comment by ljwolf** [comment]
Huh, good catch. That's probably because sparse matrices implement `**` as matrix exponentiation. For example, look at `w.sparse**0`. While correct, this is not the element-wise power that `numpy.ndarray` uses :/

This arises (iirc) because scipy sparse matrices are patterned after the [deprecated `numpy.matrix`](https://numpy.org/doc/1.18/reference/generated/numpy.matrix.html).

We just want to avoid iterate over the `w.id_order` if we can and do so using the [sparse matrix math methods](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.csc_matrix.html#scipy.sparse.csc_matrix).

I think this works how you expect:
```python
w_squared = numpy.asarray(w.sparse
                           .multiply(w.sparse)
                           .sum(axis=1)
                          ).flatten()
```

**Comment by jeffcsauer** [comment]
Ah, matrix exponentiation! Of course. Thanks again for the explanation and back-and-forth on this. I've updated the function so it no longer uses list comprehension. Addressed in commit [ed959b8](https://github.com/jeffcsauer/GSOC2020/commit/ed959b856f35921e379293c9a95b47243eeef180).
